src/eval/stack.zig

```diff
@@ -0,0 +1,250 @@
+//! and everything is effectively wiped out.
+//!
+//! When it's a function being called, or an expression that's going to evaluate to
+//! a value that has its own scope (e.g. `x = { ... }`), it's the job of the caller
+//! (or the code preceding the expression) to allocate enough space for the return value
+//! before entering the scope, and to hand a pointer to that space to the scope (or function).
+//! That way, the return value can be written to a location where the caller can access it
+//! even after the stack has been reset due to the scope ending.
+
+const std = @import("std");
+const builtin = @import("builtin");
+const testing = std.testing;
+
+/// Error when an alloca is attempted that's too big for the stack
+pub const StackOverflow = error{
+    StackOverflow,
+};
+
+/// Fixed-size stack memory allocator to be used when evaluating Roc IR
+pub const Stack = struct {
+    allocator: std.mem.Allocator,
+
+    /// The original allocation
+    start: [*]u8,
+
+    /// We only allow u32 for these, both because the stack is never going to be
+    /// so big that they would overflow, but also because we don't want compile-time
+    /// evaluation of constants to succeed on 64-bit targets and fail on 32-bit targets
+    /// (e.g. when compiling on wasm32). It should consistently work (or not) on both!
+    capacity: u32,
+
+    /// How many bytes are currently in use on the stack.
+    //
+    // If this is 0, self.next() will return self.start
+    used: u32,
+
+    pub fn initCapacity(allocator: std.mem.Allocator, capacity: u32) std.mem.Allocator.Error!Stack {
+        const memory = try allocator.alloc(u8, capacity);
+        return .{
+            .allocator = allocator,
+            .start = memory.ptr,
+            .capacity = capacity,
+            .used = 0,
+        };
+    }
+
+    pub fn deinit(self: *Stack) void {
+        self.allocator.free(self.start[0..self.capacity]);
+    }
+
+    /// Advance the stack pointer.
+    ///
+    /// This is called `alloca` because that's what libc calls this operation.
+    pub fn alloca(self: *Stack, bytes: u32, alignment: u32) StackOverflow!*anyopaque {
+        if (bytes == 0) return self.next();
+
+        // Calculate padding needed for alignment
+        const current_addr = @intFromPtr(self.next());
+        const aligned_addr = std.mem.alignForward(usize, current_addr, alignment);
+        const padding = @as(u32, @intCast(aligned_addr - current_addr));
+
+        // Check for overflow when adding padding
+        const bytes_with_padding = @addWithOverflow(bytes, padding);
+        if (bytes_with_padding[1] != 0) {
+            return StackOverflow.StackOverflow;
+        }
+
+        // Check for overflow when adding to used
+        const new_used = @addWithOverflow(self.used, bytes_with_padding[0]);
+        if (new_used[1] != 0 or new_used[0] > self.capacity) {
+            return StackOverflow.StackOverflow;
+        }
```

Comment:
nit: use tuple destructuring
```suggestion
        // Check for overflow when adding padding
        const bytes_with_padding, var overflowed = @addWithOverflow(bytes, padding);
        if (overflowed != 0) {
            return StackOverflow.StackOverflow;
        }

        // Check for overflow when adding to used
        const new_used, overflowed = @addWithOverflow(self.used, bytes_with_padding);
        if (overflowed != 0 or new_used > self.capacity) {
            return StackOverflow.StackOverflow;
        }
```